### PR TITLE
Fix Redis reconnection logic

### PR DESCRIPTION
**Pull request by erikjohnston**
This does two things:

~~1. Ensures we only send a `REPLICATE` when we (re)connect *after* we successfully subscribe. This ensures we are subscribed to hear responses to `REPLICATE` (in practice this is unlikely to be an issue as we'll almost certainly finish subscribing by the time other instances have received and responded to the `REPLICATE`).~~  Actually, this is spurious. 
2. Proactively send out `POSITION` commands (as if we had just received a `REPLICATE`) when we connect to Redis. This is important as other instances won't notice we've connected to issue a `REPLICATE` command (unlike for direct TCP connections). This is only currently an issue if master process reconnects without restarting (if it restarts then it won't have written anything and so other instances probably won't have missed anything).

### CODE REVIEW

**Comment by clokep** [comment]
This seems to try to fix the same thing as #7427 (specifically d78265af0c77b7d75bc5fd013515b53b01a2ac23)?

File: synapse/replication/tcp/handler.py [hunk 1]
```diff
@@ -151,6 +151,13 @@ def start_replication(self, hs):
             hs.get_reactor().connectTCP(host, port, self._factory)
 
     async def on_REPLICATE(self, conn: AbstractConnection, cmd: ReplicateCommand):
+        self.send_positions_to_connection(conn)
+
+    def send_positions_to_connection(self, conn: AbstractConnection):
```
Comment (clokep) [inline synapse/replication/tcp/handler.py:156]: This doesn't seem to use `conn` at all. Is that on purpose?

Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:156]: Oh, that's exciting! It *should* only be sending stuff to the given connection via `conn.send_command` but instead is sending it to all connections via `self.send_command`. It doesn't actually matter that we send `POSITION` spuriously to random connections, but lets fix it

**Comment by richvdh** [comment]
some notes for the record:

I didn't think this was necessary when I did #7427, since each process will send out a `REPLICATE` on connection, which will get echoed back and will lead to a `POSITION`. On the other hand that does feel a bit brittle/icky, so maybe this is better.

**Pull request merged**